# BW \#124 NATO Spending
NATO (North Atlantic Treaty Organization) was established after WWII in no small part to protect Western democracies against Soviet Union and other Communist threats

## Data and six questions
This week, we'll look at the 2024 NATO data on its "Defense expenditures and NATO's 2% guideline"

## Challenges
The learning goals include working with Excel files, regular expressions, cleaning data, window functions and plotting.

- Read the "defense expenditure" data from the "table 2" into a Pandas df. We want the first table on that sheet, with current prices and exchange rates. Which 5 countries have increased their NATO Spending most as a percentage, in the latest budget cycle (from 2023-2024)? Which increased most since 2014, when the data was first tracked?
- The US spends more on defense than any other NATO country, with Germany and UK in the 2nd and 3rd position. Find how much the US spent (estimated) in 2014. Then see how many other NATO countries defense spending, starting with Germany and the UK and continuing from largest to smallest, woul be needed to match US defense spending. That is, if we'd like to say that "the US spends as much as the nnext-largest contributions combined" what is n?

In [1]:
import pandas as pd

## Read the "defense expenditure" data from the "table 2" into a Pandas df. We want the first table on that sheet, with current prices and exchange rates. 

In [118]:
filename = r"C:\Users\npigeon1\Pandas-Bamboo-Weekly\BW #124 NATO Spending\240617-def-exp-2024-TABLES-en.xlsx"
df = pd.read_excel(filename, 
                   sheet_name="TABLE2",
                   skiprows=9,
                   nrows=33,
                   index_col=0,
                   usecols="C:N",
                   names=range(2013,2025)
                   )
df

,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024
2013,,,,,,,,,,,
Albania,178.358984,132.375585,130.909626,144.774189,175.894125,1.972258e+02,1.968641e+02,2.235401e+02,2.306731e+02,3.972731e+02,5.161140e+02
Belgium,5200.064956,4204.445587,4258.107663,4440.721926,4844.943851,4.761273e+03,5.323523e+03,6.244570e+03,6.903780e+03,7.622160e+03,8.519169e+03
Bulgaria,747.227377,632.718699,670.615715,724.464432,962.233571,2.158590e+03,1.120526e+03,1.275924e+03,1.439781e+03,1.991712e+03,2.325094e+03
Canada,18172.422383,18689.407920,17708.229136,23699.964684,22398.895822,2.257216e+04,2.332992e+04,2.550196e+04,2.589762e+04,2.799090e+04,3.049523e+04
Croatia,1063.800459,882.725898,836.886265,926.378076,966.071522,1.001223e+03,9.826479e+02,1.361487e+03,1.284511e+03,1.441114e+03,1.624034e+03
Czechia*,1975.310964,1921.395231,1865.735891,2258.924431,2749.767482,2.981518e+03,3.199346e+03,3.914702e+03,3.894709e+03,4.538021e+03,6.834290e+03
Denmark,4056.872333,3363.981373,3593.404282,3780.323741,4558.789015,4.487449e+03,4.885597e+03,5.274398e+03,5.472784e+03,8.140265e+03,9.939607e+03
Estonia*,513.526740,463.426697,497.433935,541.194927,615.222822,6.370754e+02,7.194874e+02,7.492753e+02,8.196315e+02,1.237741e+03,1.437315e+03
Finland,3991.243690,3400.854798,3417.744144,3535.542491,3825.481730,3.899559e+03,4.156234e+03,4.145355e+03,4.726225e+03,6.266376e+03,7.307871e+03


## Which 5 countries have increased their NATO Spending most as a percentage, in the latest budget cycle (from 2023-2024)? 

In [93]:
df.pct_change(axis='columns')[2024].nlargest()

2013
Romania*      0.542354
Czechia*      0.506007
Montenegro    0.414590
Türkiye       0.370916
Sweden        0.363539
Name: 2024, dtype: float64

In other words Romania spent 50 percent more on defense in 2024 than in 2023

## Which increased most since 2014, when the data was first tracked?

 "Let's compare each column with the one n columns before it." We can do that with the periods keyword argument, which has a default value of 1, but can be set to anything.

In [103]:
df.pct_change(axis="columns", periods=len(df.columns)-1)[2024].nlargest()

2013
Lithuania*    4.378801
Latvia*       3.838939
Hungary       3.040984
Poland*       2.460544
Czechia*      2.459855
Name: 2024, dtype: float64

## The US spends more on defense than any other NATO country, with Germany and UK in the 2nd and 3rd position. Find how much the US spent (estimated) in 2024.

In [127]:
us_spending = df.loc["United States", 2024]
us_spending

np.float64(967707.0)

In [122]:
df[2024].sort_values(ascending=False).iloc[3:]

2013
Germany            97685.590084
United Kingdom     82107.131340
France             64271.200804
Poland*            34974.527951
Italy              34461.719985
Canada             30495.234538
Türkiye            22776.382187
Netherlands        21459.679463
Spain              21269.292206
Sweden             13427.987163
Norway             10605.517023
Denmark             9939.606827
Romania*            8644.465581
Belgium             8519.169234
Greece              7684.084679
Finland             7307.870852
Czechia*            6834.290149
Hungary             4888.789011
Portugal            4626.857865
Slovak Republic     2840.608786
Bulgaria            2325.094178
Lithuania*          2300.072169
Croatia             1624.034034
Estonia*            1437.314680
Latvia*             1421.395394
Slovenia             948.969483
Luxembourg           785.057621
Albania              516.114024
North Macedonia      352.586967
Montenegro           161.832296
Name: 2024, dtype: float64

## Then see how many other NATO countries defense spending, starting with Germany and the UK and continuing from largest to smallest, woul be needed to match US defense spending. That is, if we'd like to say that "the US spends as much as the nnext-largest contributions combined" what is n?

In [129]:
df[2024].sort_values(ascending=False).iloc[3:].cumsum()

2013
Germany             97685.590084
United Kingdom     179792.721423
France             244063.922228
Poland*            279038.450178
Italy              313500.170164
Canada             343995.404701
Türkiye            366771.786889
Netherlands        388231.466352
Spain              409500.758558
Sweden             422928.745721
Norway             433534.262744
Denmark            443473.869571
Romania*           452118.335152
Belgium            460637.504386
Greece             468321.589066
Finland            475629.459917
Czechia*           482463.750067
Hungary            487352.539078
Portugal           491979.396943
Slovak Republic    494820.005728
Bulgaria           497145.099906
Lithuania*         499445.172075
Croatia            501069.206109
Estonia*           502506.520789
Latvia*            503927.916183
Slovenia           504876.885666
Luxembourg         505661.943287
Albania            506178.057311
North Macedonia    506530.644278
Montenegro         506692.476573
Name:

La somme cumulée est de 506692 < 967707. Which mean that the US defense spending is greater than all of the other NATO members put together. 

In [130]:
df[2024].sort_values(ascending=False).iloc[3:].expanding().sum().loc[lambda s_: s_ >= us_spending]

Series([], Name: 2024, dtype: float64)